server/blob/blob-deduplicating/src/main/scala/GC.scala

```diff
@@ -0,0 +1,196 @@
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.server.blob.deduplication
+
+import org.apache.james.blob.api.BlobId
+
+
+sealed abstract class Generation extends Comparable[Generation] {
+  def previous: Generation
+  def previous(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.previous)
+
+  def next: Generation
+  def next(times: Long): Generation =
+    (0L until times).foldLeft(this)((generation, _) => generation.next)
+
+  def collectibles(targetedGeneration: Generation): Set[Generation]
+  def asLong: Long
+
+  override def compareTo(that: Generation): Int = asLong.compareTo(that.asLong)
+
+  def <=(that : Generation): Boolean = compareTo(that) <= 0
+  def >(that : Generation): Boolean = compareTo(that) > 0
+}
+
+case class ValidGeneration(id: Long) extends Generation {
+  override def previous: Generation = {
+    if (id == 0) {
+      InvalidGeneration
+    } else {
+      copy(id - 1)
+    }
+  }
+  override def next: Generation = copy(id + 1)
+
+  override def collectibles(targetedGeneration: Generation): Set[Generation] =
+    ((id + 1) to targetedGeneration.previous.previous.asLong).toSet.map(Generation(_))
+
+  override def asLong: Long = id
+}
+
+object InvalidGeneration extends Generation {
+  override def previous: Generation = InvalidGeneration
+
+  override def next: Generation = ValidGeneration(0)
+
+  override def collectibles(targetedGeneration: Generation): Set[Generation] =
+    (0L to targetedGeneration.previous(GC.temporization).asLong).toSet.map(Generation(_))
+
+  override def asLong: Long = -1
+}
+
+object Generation {
+  implicit val ord: Ordering[Generation] = Ordering.by(_.asLong)
+
+  val first: Generation = apply(0)
+
+  def apply(id: Long): Generation = {
+    if (id < 0) {
+      InvalidGeneration
+    } else {
+      ValidGeneration(id)
+    }
+  }
+}
+
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(InvalidGeneration), InvalidGeneration)
```

Comment:
I doubt we want to initialize our iteration we something that is called `Invalid`